<a href="https://colab.research.google.com/github/Lolo-17/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
# Importing libraries
import pandas as pd
import numpy as np
from scipy import stats
import matplotlib.pyplot as plt

In [ ]:
# Getting file from uci
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data

--2020-08-10 23:38:12--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data.4’

imports-85.data.4   100%[===================>]  25.33K  --.-KB/s    in 0.06s   

2020-08-10 23:38:13 (400 KB/s) - ‘imports-85.data.4’ saved [25936/25936]



In [ ]:
# Reading the csv file from uci and adding columns
columns = ['symboling', 'normalized-losses', 'make', 'fuel-type', 
           'aspiration', 'num-of-doors', 'body-style', 'drive-wheels',
           'engine-location', 'wheel-base', 'length', 'width', 'height',
           'curb-weight', 'engine-type', 'num-of-cylinders', 
           'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 
           'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']

df = pd.read_csv('imports-85.data', names=columns)
print(df.shape)
df.head()

In [ ]:
# Looking at the last 5 rows
df.tail()

In [ ]:
# Checking nan's values in the dataframe
df.isnull().sum()

In [ ]:
# Checking the missing variable
df['normalized-losses'].value_counts()
# '?' is the missing variable

In [ ]:
# Changed the missing variables from '?' to NaN's
df = df.replace({'?':np.NaN})
df['normalized-losses'].value_counts(dropna=False)

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and use a 1-sample t-test to compare your estimated sample mean to the population mean. Can you say that they are different? 

**The sample is reflective to highway-mpg. Because, the population is big enough and is averaging 30. And, the sample mean is close to population mean.**

## Null Hypothesis: 

The sample mean is equal to population mean.

## Alternative Hypothesis: 

The sample mean is not equal to population mean.

Set to .5 for both.

## Confidence Level: **95%**


In [ ]:
# Gathering a sample of 10 cars with a random state of 30
df2 = df.sample(10, random_state=30)
df2.head(10)

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056
103          0               108         nissan  ...       19          25  13499
26           1               148          dodge  ...       31          38   7609
20           0                81      chevrolet  ...       38          43   6575
155          0                91         toyota  ...       27          32   8778
185          2                94     volkswagen  ...       27          34   8195

[10 rows x 26 columns]

In [ ]:
# Original population mean value
df['highway-mpg'].mean()

30.75121951219512

In [ ]:
# Sample data mean value
df2['highway-mpg'].mean()

32.8

In [ ]:
# Running a 1 sample t test
stats.ttest_1samp(df2['highway-mpg'], .5)

Ttest_1sampResult(statistic=14.050858301455087, pvalue=1.98756047234675e-07)

## Conclusion:

I reject the null hypothesis. Due to statistics of 14.14 and 
p value of .000000198

P value < .05

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

In [ ]:
# Finding the average(mean) of highway-mpg to verify salesman claim
df2 = df.sample(10, random_state=30)
df2['highway-mpg'].mean()

32.8

**The mean is not close to his claim or equal to. Therefore, is invalid.**

Salesman > Sample data mean value

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, take a sample of 100 cars. Run a t-test to verify the salesman's claim with your new larger sample. Do you reach the same conclusion as you did with the sample size of 10? 

In [ ]:
df2 = df.sample(100, random_state=30)
df2['highway-mpg'].mean()

30.64

**The mean is not close to his claim or equal to. Even with a large sample data. Therefore, is invalid.**

Salesman > Sample data mean value

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

The salesman is lying about the highway-mpg claim. Even calculating the population results in a lower value.

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
# Getting file from uci
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-08-11 00:26:16--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.07s   

2020-08-11 00:26:17 (270 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [ ]:
# Reading the csv file from uci and adding columns
column_headers = ['party','handicapped-infants','water-project',
                  'budget','physician-fee-freeze', 'el-salvador-aid',
                  'religious-groups','anti-satellite-ban',
                  'aid-to-contras','mx-missile','immigration',
                  'synfuels', 'education', 'right-to-sue','crime','duty-free',
                  'south-africa']

vt = pd.read_csv('house-votes-84.data', names=column_headers)
print(vt.shape)
vt.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [ ]:
# Replacing values to integers
vt = vt.replace({'y':1, 'n':0, '?':np.NaN})
vt.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [ ]:
# Filtering data for republicans and democrats
dems = vt[vt['party'] == 'democrat']
reps = vt[vt['party'] == 'republican']

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

### **An issue that democrats support more than republicans**

### Null Hypothesis:

Democrats support for the `handicapped-infants` bill is equal to republican levels of support.

### Alternative Hypothesis:

Democrats support for the `handicapped-infants` bill is not equal to republican levels of support.

### Confidence Level: **99%**

In [ ]:
# Looking for the P-value between the two parties for handicapped-infants bill
stats.ttest_ind(dems['handicapped-infants'], reps['handicapped-infants'], nan_policy='omit')

Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)

In [ ]:
# Getting the means from both parties for handicapped-infants bill
print(dems['handicapped-infants'].mean())
print(reps['handicapped-infants'].mean())

0.47674418604651164
0.8975903614457831


## Conclusion:

I reject the null hypothesis. Due to statistics of -9.21 and 

p value of .00000000000000000161

**P value < .01**

Also, the democrats support the handicapped-infants bill more than the

republicans by a great margin.

### **An issue that republicans support more than democrats**

### Null Hypothesis:

Republican support for the `religious-groups` bill is equal to democrat levels of support.

### Alternative Hypothesis:

Republican support for the `religious-groups` bill is not equal to democrat levels of support.

### Confidence Level: **99%**


In [ ]:
# Looking for the P-value between the two parties for religious-groups bill
stats.ttest_ind(reps['religious-groups'], dems['religious-groups'], nan_policy='omit')

Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)

In [ ]:
# Getting the means from both parties for handicapped-infants bill
print(reps['religious-groups'].mean())
print(dems['religious-groups'].mean())

0.8975903614457831
0.47674418604651164


## Conclusion:

I reject the null hypothesis. Due to statistics of -9.74 and 

p value of .0000000000000000000239

**P value < .01**

Also, the republican support the religious-groups bill more than the

democrats by a great margin.

### **An issue that democrats and republicans support a bill almost equally**

### Null Hypothesis:

Democrats support for the `water-project` bill is equal to republican levels of support.

### Alternative Hypothesis:

Democrats support for the `water-project` bill is not equal to republican levels of support.

### Confidence Level: **90%**

In [ ]:
# Looking for the P-value between the two parties for water project bill
stats.ttest_ind(dems['water-project'], reps['water-project'], nan_policy='omit')

Ttest_indResult(statistic=-0.08896538137868286, pvalue=0.9291556823993485)

In [ ]:
# Getting the means from both parties for water-project bill
print(reps['water-project'].mean())
print(dems['water-project'].mean())

0.5067567567567568
0.502092050209205


## Conclusion:

I failed to reject the null hypothesis. Due to statistics of -.09 and 

p value of .93

**P value >= .1**

Also, both parties almost support the bill equally. The republicans has

more votes than democrats however.

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE